src/palindrom-network-channel.js

```diff
@@ -0,0 +1,365 @@
+        this.palindrom = palindrom;
+
+        if (typeof window !== 'undefined' && window.location) {
+            this.remoteUrl = new URL(remoteUrl, window.location.href);
+        } else {
+            // in Node, URL is absolute
+            this.remoteUrl = new URL(remoteUrl);
+        }
+
+        onReceive && (this.onReceive = onReceive);
+        onSend && (this.onSend = onSend);
+        onConnectionError && (this.onConnectionError = onConnectionError);
+        onFatalError && (this.onFatalError = onFatalError);
+        onStateChange && (this.onStateChange = onStateChange);
+        onSocketOpened && (this.onSocketOpened = onSocketOpened);
+
+        Object.defineProperty(this, 'useWebSocket', {
+            get: function() {
+                return useWebSocket;
+            },
+            set: newValue => {
+                useWebSocket = newValue;
+
+                if (newValue == false) {
+                    if (this._ws) {
+                        this._ws.onclose = function() {
+                            //overwrites the previous onclose
+                            this._ws = null;
+                        };
+                        this._ws.close();
+                    }
+                    // define wsUrl if needed
+                } else if (!this.wsUrl) {
+                    this.wsUrl = toWebSocketURL(this.remoteUrl.href);
+                }
+                return useWebSocket;
+            }
+        });
+    }
+    get useWebSocket() {
+        return this._useWebSocket;
+    }
+    set useWebSocket(newValue) {
+        this._useWebSocket = newValue;
+
+        if (newValue == false) {
+            if (this._ws) {
+                this._ws.onclose = () => {
+                    //overwrites the previous onclose
+                    this._ws = null;
+                };
+                this._ws.close();
+            }
+            // define wsUrl if needed
+        } else if (!this.wsUrl) {
+            this.wsUrl = toWebSocketURL(this.remoteUrl.href);
+        }
+        return this.useWebSocket;
+    }
+
+    async _establish(reconnectionPendingData = null) {
+        const data = await this._xhr(
+            this.remoteUrl.href + (reconnectionPendingData ? '/reconnect' : ''),
+            'application/json',
+            reconnectionPendingData
+        );
+        if (this.useWebSocket) {
+            this.webSocketUpgrade(this.onSocketOpened);
+        }
+        return data;
+    }
+
+    /**
+     * Send any text message by currently established channel
+     * @TODO: handle readyState 2-CLOSING & 3-CLOSED (tomalec)
+     * @param  {String} msg message to be sent
+     * @return {PalindromNetworkChannel}     self
+     */
+    async send(msg) {
+        // send message only if there is a working ws connection
+        if (this.useWebSocket && this._ws && this._ws.readyState === 1) {
+            this._ws.send(msg);
+            this.onSend({ data: msg, url: this._ws.url, method: 'WS' });
```

Comment:
The signature of the `onSend` API callback was changed for no apparent reason from 3 arguments to 1 object argument.